JS review

sonoilconte opened the pull request:
Hey @captbaritone 

This is a little site I started long ago as a developer portfolio site and am finally working on it again. Would you be able to review the es6 code in 
assets/js/src/fade.js
?
Any other general critiques or suggestions you might have would be appreciated too! I have babel set up and used the airbnb style guide with eslint.

sonoilconte committed:
Delete uneeded files and run gulp/babel

captbaritone commented:
I’d love to! I’ll try to find some time today. Looking forward to it.

captbaritone commented on assets/js/src/fade.js:
```diff
@@ -1,3 +1,5 @@
+// TODO: Have someone knowledgable review this JS :)
```
😀

sonoilconte commented:
Sweet! thank you!! :-D

captbaritone commented on assets/js/dist/fade.js:
```diff
@@ -1,40 +1,38 @@
+  navElement.addEventListener('click', function (e) {
+    e.preventDefault();
+    Array.from(document.getElementsByTagName('li')).forEach(function (el) {
+      return el.classList.remove('current-btn');
```
Why do you have an explicit return here? The return value from `forEach` is ignored, so it seems odd to be going out of your way to return this value.

sonoilconte replied:
In my es6 I think I was trying to be slick doing it like this in one line
`.forEach(el =>  el.classList.remove('current-btn'));`
without stopping to realize that would be an explicit return.
If I want to do it in one line like this, does that work? 
`.forEach((el) => { el.classList.remove('current-btn') });`

captbaritone replied:
Oh, I totally forgot that I was reviewing the generated code (GitHub annoyingly won't let me add inline comments to code which did not change in the PR).

Yeah, technically the one with `{}` would be more correct, but I think it's fine to just use the arrow in this case and not worry about the extraneous return.

captbaritone commented on assets/js/dist/fade.js:
```diff
@@ -1,40 +1,38 @@
+      Array.from(document.getElementsByTagName('section')).forEach(function (el) {
+        return el.classList.remove('display');
+      });
+      document.getElementById(e.target.id.slice(4)).classList.add('display');
```
To me this `slice` call is a [code smell](https://en.wikipedia.org/wiki/Code_smell). Ids should be just ids. If you need to share additional information, maybe consider a data attribute? Think about the case where some other engineer (or you in the future) want to change this id. How would you know that this fade file has an implicit dependency on the current naming scheme?

captbaritone commented on assets/js/dist/fade.js:
```diff
@@ -1,40 +1,38 @@
+  } else {
+    document.getElementsByTagName('nav')[0].classList.remove('fixed-nav');
+    document.getElementsByClassName('wrapper')[0].classList.remove('wrapper-nav-offset');
+  }
```
When I see a function like this where two blocks are very similar, I try to identify which parts are actually unique. That can help make it easier to understand what's actually important for future readers.

Also, when I notice that two variables/classes always change together, it makes me wonder: Should we just merge the two variables? Or maybe introduce a third variable? In this case, maybe you should have a single "sticky" class on `<body>` which does both things.

Scroll events can fire very rapidly. It's good to think about both the complexity of the work you are doing in here (traversing the entire DOM tree twice to look for nodes) and also if you are doing unnecessary work. Often you can use some kind of stateful flag to track if there is work to do or not.

```
let navIsSticky = false;
window.addEventListener('scroll', function () {
  const navShouldBeSticky = window.scrollY > mobileOffset;
  if(navShouldBeSticky == navIsSticky) {
    // Nothing to to do.
    return;
  }
  document.body.classList.toggle('sticky', navShouldBeSticky);
  navIsSticky = navShouldBeSticky;
});
```

sonoilconte replied:
Ok yes. I see I'm doing a whole lot of DOM traversing and much of the time adding or removing classes that are already added or removed. :) I will play with this.

captbaritone commented on assets/js/dist/fade.js:
```diff
@@ -1,40 +1,38 @@
 
-$(document).ready(function () {
-  var mobileOffset = 270; // Start with only work section showing
+var mobileOffset = 270; // Start with nav-work button highlighted and only work section showing
```
This is currently creating an implicit dependency between this file and whatever CSS sets the height. Is there some way you could reorganize this so that there is a single source of truth?

Perhaps you could derive this value here in JS?

captbaritone commented on assets/js/dist/fade.js:
```diff
@@ -1,40 +1,38 @@
+Array.from(navElements).forEach(function (navElement) {
+  navElement.addEventListener('click', function (e) {
+    e.preventDefault();
+    Array.from(document.getElementsByTagName('li')).forEach(function (el) {
```
This selector seems too generic. It's operating on all list items, not just list items in the header.

captbaritone reviewed (commented):
Some general feedback:

* Think about the complexity of some of these operations. Consider that `document.get[*]` methods potentially have to visit every DOM node in the tree in order to derive their results. In many cases you could find the node once outside the callback, assign it to a variable, and then use that variable for future operations.
* `document.getElementById(e.target.id) === e.target`
* Have you ruled out the possibility that the fade effect could be implemented with a CSS animation?
* The fade effect is cool, but it breaks the semantics of how links work. What if I wanted to send someone a link to your contact page? What if I wanted to open each of your pages (contact, about, portfolio) in a different tab by cmd+clicking on each one? What if instead of binding to the click events, you actually had these be links to `/#about`, `#contact` and then listened for this event? https://developer.mozilla.org/en-US/docs/Web/API/WindowEventHandlers/onhashchange

captbaritone commented:
Left some feedback, I hope it's helpful! If you decide to incorporate any of it, let me know I'd be happy to look again. Thanks for asking! This was fun.

sonoilconte commented:
`document.getElementById(e.target.id) === e.target` ... good point. That is a bit of a face-palm ...uhg :)

This is super helpful thank you. It is easy to write the DOM manipulation code without stopping to think about complexity.

sonoilconte commented:
I will look into the the `hashchange` event and css fade too ... thanks again :)

sonoilconte committed:
Put nodes in variables in order to not search DOM repeatedly
Use data attribute to match nav element with appropriate section element

sonoilconte committed:
Sticky nav changes
Use css class for body that will add sticky styles to both nav and wrapper elements
For window scroll event use stateful flag to determine if class toggle needed
